src/mongodb-migrations.coffee

```diff
@@ -119,14 +120,27 @@ class Migrator
         migrationDone status: 'skip', reason: skipReason
         return runOne()
 
-      context = { db: @_db, log: userLog }
-      fn.call context, (err) ->
-        if err
+      if not @_timeout
+        timeoutProtect = true
+      else
+        timeoutProtect = setTimeout () ->
+          timeoutProtect = null
+          err = new Error "migration timed-out"
           migrationDone status: 'error', error: err
           allDone(err)
-        else
-          migrationDone status: 'ok'
-          runOne()
+        , @_timeout
+
+      context = { db: @_db, log: userLog }
+      fn.call context, (err) ->
+        if timeoutProtect
```

Comment:
I see what you did here with `true`, which is clever but a bit hard to grasp IMO.
I would split it into two variables:

`timeoutId` (only set when needed), and `isCallbackCalled` (set to `false` initially, in case of timeout set to `true` inside of the interval).

Then here you would blindly clear `timeoutId` (which will happily swallow null/undefined), then `return if isCallbackCalled`, and then have the normal handling (`if err...`) unchanged.